diff --git a/labs/advanced1/gpt2.py b/labs/advanced1/gpt2.py
index 23f6ba2..aa53a01 100644
--- a/labs/advanced1/gpt2.py
+++ b/labs/advanced1/gpt2.py
@@ -6,15 +6,18 @@ import torch.nn as nn
 
 @dataclass
 class Config:
-    n_vocab = 50257
-    n_ctx = 1024
+    n_vocab = 30522
+    n_ctx = 512
     n_embd = 768
     n_head = 12
     n_layer = 12
+    n_type_vocab_size = 2
 
 
+# def gelu(x):
+#     return 0.5 * x * (1 + torch.tanh((2 / torch.pi) ** 0.5 * (x + 0.044715 * x**3)))
 def gelu(x):
-    return 0.5 * x * (1 + torch.tanh((2 / torch.pi) ** 0.5 * (x + 0.044715 * x**3)))
+    return x * 0.5 * (1.0 + torch.erf(x / (2.0 ** 0.5)))
 
 
 class MLP(nn.Module):
@@ -31,10 +34,6 @@ class MLP(nn.Module):
         return x
 
 
-def make_causal_mask(n):
-    return torch.triu(torch.full((n, n), float("-inf")), diagonal=1)
-
-
 class Attention(nn.Module):
     def __init__(self, config):
         super().__init__()
@@ -42,9 +41,9 @@ class Attention(nn.Module):
         self.n_head = config.n_head
         self.c_attn = nn.Linear(config.n_embd, config.n_embd * 3)
         self.c_proj = nn.Linear(config.n_embd, config.n_embd)
-        self.register_buffer("mask", make_causal_mask(config.n_ctx), persistent=False)
+        # self.register_buffer("mask", make_causal_mask(config.n_ctx), persistent=False)
 
-    def forward(self, x):
+    def forward(self, x, attention_mask):
         batch_size, seq_len, n_embd = x.shape
         head_embd = n_embd // self.n_head
         q, k, v = self.c_attn(x).chunk(3, dim=-1)
@@ -56,7 +55,10 @@ class Attention(nn.Module):
         v = v.transpose(-2, -3)
         x = q @ k.transpose(-1, -2)
         x = x / head_embd**0.5
-        x = x + self.mask[:seq_len, :seq_len]
+        if attention_mask is not None:
+            # [B, S] -> [B, 1, 1, S], 1 keeps token, 0 masks token
+            key_mask = attention_mask[:, None, None, :].to(dtype=torch.bool)
+            x = x.masked_fill(~key_mask, torch.finfo(x.dtype).min)
         x = torch.softmax(x, dim=-1)
         x = x @ v
         x = x.transpose(-2, -3).contiguous()
@@ -74,7 +76,7 @@ class LayerNorm(nn.Module):
     def forward(self, x):
         mean = x.mean(dim=-1, keepdim=True)
         variance = x.var(unbiased=False, dim=-1, keepdim=True)
-        return self.g * (x - mean) / torch.sqrt(variance + 1e-05) + self.b
+        return self.g * (x - mean) / torch.sqrt(variance + 1e-12) + self.b
 
 
 class Block(nn.Module):
@@ -85,9 +87,9 @@ class Block(nn.Module):
         self.ln_2 = LayerNorm(config)
         self.mlp = MLP(config)
 
-    def forward(self, x):
-        x = x + self.attn(self.ln_1(x))
-        x = x + self.mlp(self.ln_2(x))
+    def forward(self, x, attention_mask):
+        x = self.ln_1(x+self.attn(x, attention_mask))
+        x = self.ln_2(x+self.mlp(x))
         return x
 
 
@@ -101,17 +103,30 @@ class Model(nn.Module):
         self.config = config
         self.wte = nn.Embedding(config.n_vocab, config.n_embd)
         self.wpe = nn.Embedding(config.n_ctx, config.n_embd)
-        self.h = nn.Sequential(*(Block(config) for _ in range(config.n_layer)))
+        self.type_token_wpe = nn.Embedding(config.n_type_vocab_size, config.n_embd)
+        self.h = nn.ModuleList([Block(config) for _ in range(config.n_layer)])
         self.ln_f = LayerNorm(config)
-        self.lm_head = nn.Linear(config.n_embd, config.n_vocab, bias=False)
+        # self.lm_head = nn.Linear(config.n_embd, config.n_vocab, bias=False)
+        self.cls_pooler = nn.Linear(config.n_embd, config.n_embd)
+        self.cls_pooler_activation = nn.Tanh()
         self.register_buffer("pos", make_positions(config.n_ctx), persistent=False)
 
-    def forward(self, x):
+    def forward(self, x=None, token_type_ids=None, attention_mask=None, input_ids=None):
+        if x is None:
+            x = input_ids
         batch_size, seq_len = x.shape
         wte = self.wte(x)
         wpe = self.wpe(self.pos[:seq_len])
-        x = wte + wpe
-        x = self.h(x)
+
+        if attention_mask is None:
+            attention_mask = torch.ones_like(x)
+        if token_type_ids is None:
+            token_type_ids = torch.zeros_like(x)
+
+        type_token_wte = self.type_token_wpe(token_type_ids)
+        x = wte + wpe + type_token_wte
         x = self.ln_f(x)
-        x = self.lm_head(x)
-        return x
+        for h in self.h:
+            x = h(x, attention_mask=attention_mask)
+        pooled = self.cls_pooler_activation(self.cls_pooler(x[:, 0]))
+        return (x, pooled)
